### PR TITLE
Add cloud functions module

### PR DESCRIPTION
**Pull request by ckeun**
Generates two cloud functions: 
(1) getMessages - retrieve messages by chain with the option of filtering for messages with missing vaas
(2) getMessageCounts - retrieve total counts across wormhole chains along with total counts of messages with missing vaas

### CODE REVIEW

**Comment by ckeun** [comment]
With the getRows ranges implementation: 
initial load time for getMessageCounts is 19 sec and each cache refresh is 600 ms
initial load time for getMessages is 4 sec (for Eth) and each cache refresh is 800 ms. however, cache refresh for hasSignedVaa=0 takes longer b/c you have to go back more blocks to find enough messages to return 😬

**Comment by ckeun** [comment]
In getMessageCounts, realized that the numMessagesWithoutVaas metric wasn't updating correctly. Needed to do a fresh pull of rows to get updated data. Now storing the firstMissingVaaRowKey to have a starting point for those row ranges. 
The process now does two getRows() calls: (1) gets new messages, (2) gets missing vaas). This increases the overall time, but once we backfill the older missing vaas, the overall pull time should be ~1-1.5 sec.

File: cloud_functions/scripts/deploy.sh [hunk 1]
```diff
@@ -0,0 +1,27 @@
...
+
+if [ -z "$BIGTABLE_TABLE_ID" ]; then
+    echo "BIGTABLE_TABLE_ID must be specified"
+    exit 1
+fi
+
+# note CLOUD_FUNCTIONS_NUM_ROWS isn't required and defaults to 100 if not provided
+
+# for initial deployment
+gcloud functions deploy messages-test  --entry-point getMessages  --runtime nodejs16 --trigger-http --allow-unauthenticated --timeout 300 --memory 1GB
```
Comment (ckeun) [inline cloud_functions/scripts/deploy.sh:18]: add flag for region "--region". deploy in same region as wormhole-mainnet "europe-west-3"

File: cloud_functions/src/getMessageCounts.ts [hunk 2]
```diff
@@ -0,0 +1,114 @@
...
+    ranges: [],
+  };
+
+  // build range values for each chain based on first missing vaa row key
+  let missingVaaRanges: { ranges: { start: string; end: string }[] } = {
+    ranges: [],
+  };
+
+  for (const [prevChain, prevData] of Object.entries(prevMessageCounts)) {
+    const startRowKey = prevData.lastRowKey || padUint16(Number(prevChain).toString());
```
Comment (evan-gray) [inline cloud_functions/src/getMessageCounts.ts:30]: We'll need a way to reset these for when we fill gaps. Is it possible to add a flag like `forceRefresh` or something that we plan to call manually?

File: cloud_functions/src/getMessageCounts.ts [hunk 3]
```diff
@@ -0,0 +1,114 @@
...
+    Object.keys(cache['messages']).length === 0 ||
+    Date.now() - cache['lastUpdated'] > REFRESH_TIME_INTERVAL
+  ) {
+    if (Object.keys(cache['messages']).length === 0) {
+      console.log(`cache is empty, setting cache['messages] ${new Date()}`);
+    } else {
+      console.log(`cache is older than ${REFRESH_TIME_INTERVAL} ms, refreshing ${new Date()}`);
+    }
+    let prevDate = Date.now();
+    messages = await getMessageCounts_(cache['messages']);
```
Comment (evan-gray) [inline cloud_functions/src/getMessageCounts.ts:103]: I think you can simply try catch here and in the catch do a console.error, res.sendStatus(500), and return

File: cloud_functions/src/getMessageCounts.ts [hunk 4]
```diff
@@ -0,0 +1,114 @@
...
+      prevData.firstMissingVaaRowKey || padUint16(Number(prevChain).toString());
+    const endRowKey = padUint16((Number(prevChain) + 1).toString());
+
+    // TODO: use inclusive/exclusive?
+    ranges['ranges'].push({ start: startRowKey, end: endRowKey });
+    missingVaaRanges['ranges'].push({ start: startMissingVaaRowKey, end: endRowKey });
+  }
+
+  let messageCounts: CountsByChain = {};
+  try {
```
Comment (evan-gray) [inline cloud_functions/src/getMessageCounts.ts:41]: probably can avoid the try catch here in lieu of doing it outside so you can return a 500

File: cloud_functions/src/getMessages.ts [hunk 5]
```diff
@@ -0,0 +1,453 @@
...
+          },
+        ],
+      });
+
+      const filteredMessages = observedMessages.filter((o) => o?.id !== end);
+
+      for (const message of filteredMessages.reverse().slice(0, numMessages)) {
+        messages.push({
+          id: message?.id,
+          chain: parseInt(message?.id.split('/')[0]),
```
Comment (evan-gray) [inline cloud_functions/src/getMessages.ts:230]: might want to perform this with 1 split like
```
const [chain, block, emitter, seq] = parseInt(message?.id.split('/')
```
and probably wrap this in an `if (message)` so you don't have to `?` everywhere

File: cloud_functions/src/getMessages.ts [hunk 6]
```diff
@@ -0,0 +1,453 @@
...
+  var keys: never[] = [];
+  var re = pathToRegexp('/:chain/:noVaa?', keys, {
+    strict: false,
+  });
+  var pathVars = re.exec(req.path);
+  let chain = undefined;
+  let noVaasBool = undefined;
+  if (pathVars) {
+    chain = pathVars[1];
+    // TODO: make noVaasBool a query param?
```
Comment (evan-gray) [inline cloud_functions/src/getMessages.ts:387]: I approve of this TODO

File: cloud_functions/src/getMessages.ts [hunk 7]
```diff
@@ -0,0 +1,453 @@
...
+  messages: ObservedMessage[];
+  lastUpdated: number;
+  lastRowKey: string;
+};
+
+export type MessagesByChain = {
+  [chain in ChainId]: ObservedMessageResponse;
+};
+
+let noVaaCache: MessagesByChain = {
```
Comment (evan-gray) [inline cloud_functions/src/getMessages.ts:27]: change this to a function that returns the empty object. then call it once per cache

**Review by evan-gray — CHANGES_REQUESTED** [review]
Looking at these queries, feels like we want to index by the inverse block

File: cloud_functions/src/getMessages.ts [hunk 8]
```diff
@@ -0,0 +1,290 @@
...
+  const instance = bigtable.instance(assertEnvironmentVariable('BIGTABLE_INSTANCE_ID'));
+  const table = instance.table(assertEnvironmentVariable('BIGTABLE_TABLE_ID'));
+  let messages: ObservedMessage[] = [];
+  let lastRowKey = fromKey;
+  try {
+    const prefix = `${padUint16(chainId.toString())}/`;
+    let start = fromKey === undefined ? prefix : fromKey;
+    let end = `${padUint16((chainId + 1).toString())}/`;
+    var [observedMessages] = await table.getRows({
+      limit: numMessages,
```
Comment (evan-gray) [inline cloud_functions/src/getMessages.ts:42]: if `fromKey` was passed in, can we do `numMessages+1` and then not return the first one?

**Review by evan-gray — CHANGES_REQUESTED** [review]

**Review by evan-gray — APPROVED** [review]

**Comment by evan-gray** [comment]
Great first iteration.

**Pull request merged**